hazelcast/src/main/java/com/hazelcast/sql/impl/expression/string/SubstringFunction.java

```diff
@@ -0,0 +1,115 @@
+ */
+
+package com.hazelcast.sql.impl.expression.string;
+
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.expression.TriExpression;
+import com.hazelcast.sql.impl.expression.math.MathFunctionUtils;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+
+public class SubstringFunction extends TriExpression<String> implements IdentifiedDataSerializable {
+    public SubstringFunction() {
+        // No-op
+    }
+
+    private SubstringFunction(Expression<?> input, Expression<?> start, Expression<?> length) {
+        super(input, start, length);
+    }
+
+    public static SubstringFunction create(Expression<?> input, Expression<?> start, Expression<?> length) {
+        return new SubstringFunction(input, start, length);
+    }
+
+    @SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:NPathComplexity"})
+    @Override
+    public String eval(Row row, ExpressionEvalContext context) {
+        // Get input
+        String input;
+
+        try {
+            input = StringExpressionUtils.asVarchar(operand1, row, context);
+        } catch (Exception e) {
+            // Conversion to String failed. E.g. NPE on UserClass.toString()
+            throw QueryException.dataException("Failed to get value of input operand of SUBSTRING function", e);
+        }
+
+        if (input == null) {
+            // NULL always yields NULL
+            return null;
+        }
+
+        Integer start = operand2 != null ? MathFunctionUtils.asInt(operand2, row, context) : null;
+
+        if (start == null) {
+            // Start position is not specific, start with the very first character
+            start = 1;
+        } else if (start < 1) {
+            // Different databases provide different semantics on negative values. Oracle start counting
+            // from the end, SQL Server starts from the beginning, and uses the value to calculate the
+            // final length, etc.
+            // These semantics are pretty complicated, so wi disallow it completely.
+            throw QueryException.dataException("SUBSTRING \"start\" operand must be positive");
+        }
+
+        // In SQL start position is 1-based. Convert it to 0-based for Java.
+        int adjustedStart = start - 1;
+
+        if (adjustedStart > input.length()) {
+            // Start position is beyond the string length, e.g. SUBSTRING("abc", 4)
+            return "";
+        }
+
+        Integer length = operand3 != null ? MathFunctionUtils.asInt(operand3, row, context) : null;
+
+        if (length == null) {
+            // Length is not specified, just cut from the start
+            return adjustedStart > 0 ? input.substring(adjustedStart) : input;
```

Comment:
if start was provided and its value is null, it should return null

Reply:
Fixed in https://github.com/hazelcast/hazelcast/pull/17379/commits/37f0ac6b7b8fb8fbe440400f71d934f7044ee78b